Support for color-scheme preference on Linux

l0drex opened the pull request:
Linux support for all desktops that implement the `color-scheme` preference. Implementation is based on [this gist](https://gist.github.com/DevSrSouza/b013d1a8119f50615a493b36cf0b9b56).
First time using Gradle and Kotlin, any help is welcome :smile: 

---
Fix #56
Fix #50 
Fix #33
Fix #20

l0drex commented:
I wasn't able to run it yet, it always hangs while processing resources and compiling Kotlin

weisJ commented on linux/xdg-desktop-portal/build.gradle.kts:
```diff
@@ -0,0 +1,88 @@
+plugins {
+    java
+    id("dev.nokee.jni-library")
+    id("dev.nokee.cpp-language")
+    `uber-jni-jar`
+    `use-prebuilt-binaries`
```
If you don't plan to interface with any C/C++ code you don't have to add these plugins

l0drex replied:
Makes sense, I deleted those parts

weisJ commented on linux/xdg-desktop-portal/build.gradle.kts:
```diff
@@ -0,0 +1,88 @@
+    implementation(libs.darklaf.nativeUtils)
+
+    // TODO add xdg-desktop-portal dependency
+    implementation("com.github.hypfvieh:dbus-java:3.3.1")
+    implementation("com.github.hypfview:dbus-java-transport-jnr-unixsocket:4.0.0")
```
Those dependencies don't seem to exists according to CI

l0drex replied:
My IDE can find dbus-java, but not the second one. It's in the [dbus-java repository](https://github.com/hypfvieh/dbus-java), so I just commented it out for now. However, I don't know if that works since I can't build it yet...

weisJ commented on linux/xdg-desktop-portal/build.gradle.kts:
```diff
@@ -0,0 +1,88 @@
+                        when (it) {
+                            is Gcc, is Clang -> listOf("-O2")
+                            else -> emptyList()
+                        }
+                    }
+                )
+            }
+            linkTask.configure {
+                linkerArgs.addAll(
+                    toolChain.map {
+                        when (it) {
+                            is Gcc, is Clang -> linkerFlagsFor(
+                                "glibmm-2.4",
+                                "giomm-2.4",
+                                "sigc++-2.0",
+                                "gtk+-3.0"
+                            )
+                            else -> emptyList()
+                        }
+                    }
+                )
+            }
+        }
+    }
+}
+
+fun compilerFlagsFor(vararg packages: String): List<String> =
+    "pkg-config --cflags ${packages.joinToString(separator = " ")}".runCommand().split(" ").distinct()
+
+fun linkerFlagsFor(vararg packages: String): List<String> =
+    "pkg-config --libs ${packages.joinToString(separator = " ")}".runCommand().split(" ").distinct()
+
+fun String.runCommand(): String {
+    val process = ProcessBuilder(*split(" ").toTypedArray()).start()
+    val output = process.inputStream.reader(Charsets.UTF_8).use {
+        it.readText()
+    }
+    process.waitFor(10, TimeUnit.SECONDS)
+    return output.trim()
+}
```
Again without C/C++ code this is uneccessary.

weisJ commented on linux/xdg-desktop-portal/src/main/java/com/github/weisj/darkmode/platform/linux/xdg/XdgLibrary.java:
```diff
@@ -0,0 +1,47 @@
+import com.github.weisj.darkmode.platform.LibraryUtil;
+import com.github.weisj.darkmode.platform.PluginLogger;
+
+public class XdgLibrary extends AbstractPluginLibrary {
+    private static final String PROJECT_NAME = "auto-dark-mode-linux-xdg";
+    private static final String PATH = "/com/github/weisj/darkmode/" + PROJECT_NAME + "/linux-x86-64/";
+    private static final String DLL_NAME = "lib" + PROJECT_NAME + ".so";
+    private static final XdgLibrary instance = new XdgLibrary();
+
```
No C/C++ no `AbstractPluginLibrary` necessary :). Specifically this would fail when calling `XdgLibrary.get()` as the library binary doesn't actually exist.

weisJ reviewed (commented):
Just a few comments. I'm not sure why it's spinning on `prepareResources` for you though. What command are you using for building? Also this doesn't seem to actually implement detecting changes to the `org.freedesktop.appearance/color-scheme`. Will this be added?

l0drex commented:
I can implement detecting changes, there is a signal for that. For now, I just wanted basic functionality and worry about that once I know the concept works.

To test the implementation, I wrote a test under .../xdg-desktop-portal/src/test/... and run it using the green arrow that IDEA shows next to the test. I also tried to write scratch files, but that didn't work either. Please let me know if there is a better way :sweat_smile:

weisJ commented:
> To test the implementation, I wrote a test under .../xdg-desktop-portal/src/test/... and run it using the green arrow that IDEA shows next to the test. I also tried to write scratch files, but that didn't work either. Please let me know if there is a better way 😅

You'll have to setup the gradle project to use junit.

````kotlin
depenencies {
...
testImplementation(libs.test.junit.api)
testRuntimeOnly(libs.test.junit.engine)
}

tasks  {
    test {
        testLogging {
            exceptionFormat = org.gradle.api.tasks.testing.logging.TestExceptionFormat.FULL
            showStandardStreams = true
        }
        useJUnitPlatform()
    }
}
````
IntelliJ should automatically pick up the gradel task to run tests. If not use `gradlew :auto-dark-mode-linux-xdg-desktop-portal:test`

If you run into the build spinning on `prepareResources` you can try to run `gradlew build [whatever command you wanted to execute]`. It seems like there are same issues with task ordering where multiple tasks wait on each other causing the dead-lock. I'll have to investigate what is causing it.

l0drex commented:
Thanks a lot, its now working! I had to update the Kotlin version to `1.6.0` in `gradle.properties`, I hope thats not a problem. Without that, gradle complained that some manifests were using incompatible versions.

I now have some questions on the implementation:
1. Should `isDarkThemeEnabled` throw an exception if its not supported (or otherwise does not know what theme is active) or should it rather return dark / light as a fallback?
2. There is no way to detect high contrast themes yet, should I leave that TODO as is or return a value?

weisJ commented:
> Thanks a lot, its now working!
> 
> I now have some questions on the implementation:
> 
> 1. Should `isDarkThemeEnabled` throw an exception if its not supported (or otherwise does not know what theme is active) or should it rather return dark / light as a fallback?

In this case it should just return the current state i.e. not result in a change at all.

> 2. There is no way to detect high contrast themes yet, should I leave that TODO as is or return a value?

Please simply return `false` in this case.

l0drex commented:
I have some issues implementing the event handling. I guess `createEventHandler` and `deleteEventHandler` both implement an observer pattern and are called to register or remove event handlers. The creator returns a pointer to the registered callback, whereas the delete method points to the callback that should be deleted.
My issue is the `NativePointer` type. From what I understand, these are used together with C/C++ code, but I don't have that.
I found a way to cast the native pointer in the delete method to a callable:
```Kotlin
val callback = eventHandle.castSafelyTo<() -> Unit>()
```

However, I couldn't find a way to cast a callable to a native pointer...

Another thing I couldn't find was a way to return the currently used theme mode of the IDE. Is there a boolean that stores the current theme mode, or can I check if the currently used theme is dark?

weisJ commented:
> I have some issues implementing the event handling. I guess `createEventHandler` and `deleteEventHandler` both implement an observer pattern and are called to register or remove event handlers. The creator returns a pointer to the registered callback, whereas the delete method points to the callback that should be deleted. My issue is the `NativePointer` type. From what I understand, these are used together with C/C++ code, but I don't have that.
> 

The contract of the class is that `createEventHandler` and `deleteEventHandler` are only called once (and after each other).
In `createEventHandler` you can simply initialize the listener and return any `NativePointer` e.g. `NativePointer(0)`. Then in
`deleteEventHandler` just revert whatever happened during initialization. No need for casting.

> However, I couldn't find a way to cast a callable to a native pointer...
> 

Because it isn't possible ;)

> Another thing I couldn't find was a way to return the currently used theme mode of the IDE. Is there a boolean that stores the current theme mode, or can I check if the currently used theme is dark?

There are two options here. Either keep track of the current state yourself and return it, or change the type of callback from `() -> Unit` to `ThemeInfo -> Unit`, where
````kotlin
data class ThemeInfo(val isDark, val isHighContrast)
````
You'll have to adjust the interface of `ThemeMonitorService` and implementations/usage accordingly.

l0drex commented:
I looked a bit deeper through the code and I found that if `isSupported` is `false`, an exception is thrown and `isDarkThemeEnabled` is never called, so the value of that when not supported doesn't matter I guess.

weisJ commented on linux/xdg/build.gradle.kts:
```diff
@@ -0,0 +1,17 @@
+
+    kapt(libs.autoservice.processor)
+    compileOnly(libs.autoservice.annotations)
+    compileOnly("org.jetbrains.kotlin:kotlin-stdlib-jdk8:1.6.20")
```
```suggestion
    compileOnly(kotlin("stdlib-jdk8"))
```

weisJ commented on linux/xdg/build.gradle.kts:
```diff
@@ -0,0 +1,17 @@
+plugins {
+    java
+    kotlin("jvm")
+    kotlin("kapt")
```
Please use ksp instead of kapt.
```suggestion
    id("com.google.devtools.ksp")
```

weisJ commented on linux/xdg/build.gradle.kts:
```diff
@@ -0,0 +1,17 @@
+    implementation("com.github.hypfvieh:dbus-java:3.3.1")
+    implementation("org.jetbrains.kotlinx:kotlinx-coroutines-core-jvm:1.6.1")
+
+    kapt(libs.autoservice.processor)
```
```suggestion
    ksp(libs.autoservice.processor)
```

weisJ commented on linux/xdg/build.gradle.kts:
```diff
@@ -0,0 +1,17 @@
+    implementation(projects.autoDarkModeBase)
+    implementation(libs.darklaf.nativeUtils)
+
+    implementation("com.github.hypfvieh:dbus-java:3.3.1")
+    implementation("org.jetbrains.kotlinx:kotlinx-coroutines-core-jvm:1.6.1")
+
```
Please declare those in https://github.com/weisJ/auto-dark-mode/blob/master/gradle/libs.versions.toml and use the type safe accessors.

weisJ commented on linux/xdg/build.gradle.kts:
```diff
@@ -0,0 +1,17 @@
+
+dependencies {
+    implementation(projects.autoDarkModeBase)
+    implementation(libs.darklaf.nativeUtils)
```
This dependency is unused and can be removed.

weisJ commented on linux/xdg/src/main/java/com/github/weisj/darkmode/platform/linux/xdg/XdgThemeMonitorService.kt:
```diff
@@ -0,0 +1,80 @@
+ *  2: light
+ */
+class XdgThemeMonitorService : ThemeMonitorService, DBusSigHandler<FreedesktopInterface.SettingChanged> {
+    private val connection = DBusConnection.getConnection(DBusConnection.DBusBusType.SESSION)
+    private val freedesktopInterface: FreedesktopInterface = connection.getRemoteObject(
+        "org.freedesktop.portal.Desktop",
+        "/org/freedesktop/portal/desktop",
+        FreedesktopInterface::class.java
+    )
```
Can this be inside the companion object? i.e.
````kotlin
object {
    private val CONNECTION = ...
    private val FREEDESKTOP_INTERFACE = ...
}
````

weisJ commented on linux/xdg/src/main/java/com/github/weisj/darkmode/platform/linux/xdg/FreedesktopInterface.kt:
```diff
@@ -0,0 +1,17 @@
+    class SettingChanged(objectpath: String, namespace: String, key: String, value: Variant<Any>) :
+        DBusSignal(objectpath, namespace, key, value) {
+        val colorSchemeChanged: Boolean =
+            namespace == "org.freedesktop.appearance" && key == "color-scheme"
```
Those constants are repeated in `XdgThemeMonitorService`. Please make them constants in `FreeDesktopInterface`
````kotlin
object {
    const val APPEARANCE_NAMESPACE = "org.freedesktop.appearance"
    const val COLOR_SHEME_KEY = "color-scheme"
}
````

weisJ commented on linux/xdg/src/main/java/com/github/weisj/darkmode/platform/linux/xdg/XdgThemeMonitorService.kt:
```diff
@@ -0,0 +1,80 @@
+    )
+    private var eventHandler: (() -> Unit)? = null
+
+    private val themeMode: Number
+        get() {
+            val theme = freedesktopInterface.Read("org.freedesktop.appearance", "color-scheme")
+            return recursiveVariantValue(theme) as Int
+        }
```
I think it would be clearer if this could return an enum with values `LIGHT, DARK, ERROR`

weisJ commented on linux/xdg/src/main/java/com/github/weisj/darkmode/platform/linux/xdg/XdgThemeMonitorService.kt:
```diff
@@ -0,0 +1,80 @@
+        }
+
+    override val isDarkThemeEnabled: Boolean
+        get() {
+            return themeMode != 2
+        }
```
This can be inlined. Same applies to `isHighContrastEnabled` and `isSupported`.
```suggestion
        get() = themeMode != 2
```

weisJ commented on linux/xdg/src/main/java/com/github/weisj/darkmode/platform/linux/xdg/XdgThemeMonitorService.kt:
```diff
@@ -0,0 +1,80 @@
+        }
+
+    init {
+        connection.addSigHandler(FreedesktopInterface.SettingChanged::class.java, this)
```
Please register this in `createEventHandler` and unregister it in `deleteEventHandler`. Otherwise the plugin won't be able to be unloaded without restart as the reference will keep the class instance alive.

weisJ commented on linux/xdg/src/main/java/com/github/weisj/darkmode/platform/linux/xdg/XdgThemeMonitorService.kt:
```diff
@@ -0,0 +1,80 @@
+    }
+
+    override fun createEventHandler(callback: () -> Unit): NativePointer? {
+        if (eventHandler != null) {
+            throw IllegalStateException("There is already an eventhandler initialized.")
+        }
```
```suggestion
    checkNotNull(eventHandler) { "Event handler already initialized" }
```

l0drex replied:
Should be `check(eventHandler == null) ...`

zliuva commented:
> I wasn't able to run it yet, it always hangs while processing resources and compiling Kotlin

I ran into a similar issue and the solution for me was to disable parallel builds, in `gradle.properties`, set `org.gradle.parallel = false` (from true)

weisJ commented:
`XdgThemeMonitorService` still isn't used at all. You'll have to return it in https://github.com/weisJ/auto-dark-mode/blob/master/linux/src/main/java/com/github/weisj/darkmode/platform/linux/LinuxThemeMonitorService.kt
I would advise to check whether `isSupported` is true in there. Please don't remove the Gnome implementation, but use the xdg one as the alternative.

l0drex commented on linux/build.gradle.kts:
```diff
@@ -15,5 +15,6 @@ dependencies {
      * will be pulled into plugin/build.gradle.kts when it specifies implementation(project(":auto-dark-mode-linux")).
      */
     implementation(project(":auto-dark-mode-linux-gnome"))
+    implementation(project(":auto-dark-mode-linux-xdg"))
```
I tried to run the `runIde` task, but it fails because it can't find the `SigHandler` class (defined in `dbus-java`). I guess I have missed something in the gradle setup.

This is strange, because for my testing class I just added `testImplementation(projects.autoDarkModeLinuxXdg)` to the gradle setup of /plugin and everything works...

weisJ replied:
Is this a runtime error or a build error? In the case of the former make sure you have run `gradlew build` before.

l0drex replied:
It is a build error. Exact output is:
```
Supertypes of the following classes cannot be resolved. Please make sure you have the required dependencies in the classpath:
    class com.github.weisj.darkmode.platform.linux.xdg.XdgThemeMonitorService, unresolved supertypes: org.freedesktop.dbus.interfaces.DBusSigHandler
Adding -Xextended-compiler-checks argument might provide additional information.
```
I tried invalidating the caches and removed the `.gradle` directory to redownload everything, but the error remains.

weisJ replied:
You’ll have to make `XdgThemeMonitorService` not extend `DBusSigHandler`, as it isn’t exported as api (which it shouldn’t). Instead implement `DBusSigHandler` through a separate class of which `XdgThemeMonitorService` holds an instance of.

weisJ commented on linux/xdg/src/main/java/com/github/weisj/darkmode/platform/linux/xdg/FreedesktopInterface.kt:
```diff
@@ -0,0 +1,67 @@
+        const val APPEARANCE_NAMESPACE = "org.freedesktop.appearance"
+        const val COLOR_SCHEME_KEY = "color-scheme"
+
+        private val connection = DBusConnection.getConnection(DBusConnection.DBusBusType.SESSION)
+        private val freedesktopInterface: FreedesktopInterface = connection.getRemoteObject(
+            "org.freedesktop.portal.Desktop",
+            "/org/freedesktop/portal/desktop",
+            FreedesktopInterface::class.java
+        )
```
Actually thinking about this some more I'd prefer these variables to be members (same goes for the functions below), with `XdgThemeMonitorService` holding an instance of it. Otherwise we uneccessarily initialize everything in here even If it isn't used (which may also prevent the plugin being unloaded)

l0drex replied:
So do I move those back to the theme monitor service, or should I create a new class holding the connection and functions? Because I can't make them members of the interface...

weisJ replied:
Does it need to be an interface? If not just make it a class.

l0drex commented on linux/src/main/java/com/github/weisj/darkmode/platform/linux/LinuxThemeMonitorService.kt:
```diff
@@ -29,12 +29,17 @@ import com.github.weisj.darkmode.platform.Notifications
 
 class LinuxThemeMonitorService : ThemeMonitorService by if (LibraryUtil.isGnome) {
     GnomeThemeMonitorService()
+} else if (FreedesktopInterface.theme != ThemeMode.ERROR) {
+    XdgThemeMonitorService()
 } else {
```
I can't use return, so I guess you mean I should do this:
```kotlin
val xdgThemeMonitorService = XdgThemeMonitorService()
class LinuxThemeMonitorService : ThemeMonitorService by if (LibraryUtil.isGnome) {
    GnomeThemeMonitorService()
} else if (xdgThemeMonitorService.isSupported) {
    xdgThemeMonitorService
} else { ...
```

weisJ replied:
More like
```kotlin
class LinuxThemeMonitorService : ThemeMonitorService by if (LibraryUtil.isGnome) {
    GnomeThemeMonitorService()
} else { 
    val xdgThemeMonitorService = XdgThemeMonitorService()
    if (xdgThemeMonitorService.isSupported) {
        xdgThemeMonitorService
    } else { ...
}
```

l0drex commented:
There seems to be a deadlock somewhere, when I run the IDE it spins endlessly (at task `runIDE`) and never opens the window. I have disabled `org.gradle.parallel` as suggested by @zliuva which fixed the spin with resources, but not this one. I also let IDEA search for cyclic dependencies, but it couldn't find any.
I noticed that it fails to load xapp, canberry and appmenu, but from my googling that doesn't seem to be the problem.

Any suggestions? :sweat_smile:

---

<details>
<summary>Debug Log</summary>
<pre>
...
2022-05-12T18:10:35.399+0200 [DEBUG] [org.jetbrains.intellij.IntelliJPlugin] [gradle-intellij-plugin :auto-dark-mode-pluginauto-dark-mode-plugin:auto-dark-mode-plugin:runIde] Runtime specified with ideDir='/home/l0drex/.gradle/caches/modules-2/files-2.1/com.jetbrains.intellij.idea/ideaIC/2022.1/c91a19824ffcb28872ecd2761c83b250f703ff21/ideaIC-2022.1', version='11_0_14_1b2043.25' resolved as: /home/l0drex/.gradle/caches/modules-2/files-2.1/com.jetbrains/jbre/jbr_jcef-11_0_14_1-linux-x64-b2043.25/extracted/jbr/bin/java
2022-05-12T18:10:35.399+0200 [INFO] [org.jetbrains.intellij.IntelliJPlugin] [gradle-intellij-plugin :auto-dark-mode-pluginauto-dark-mode-plugin:auto-dark-mode-plugin:runIde] Resolved JVM Runtime directory: /home/l0drex/.gradle/caches/modules-2/files-2.1/com.jetbrains/jbre/jbr_jcef-11_0_14_1-linux-x64-b2043.25/extracted/jbr/bin/java
2022-05-12T18:10:35.399+0200 [INFO] [org.gradle.process.internal.DefaultExecHandle] Starting process 'command '/home/l0drex/.gradle/caches/modules-2/files-2.1/com.jetbrains/jbre/jbr_jcef-11_0_14_1-linux-x64-b2043.25/extracted/jbr/bin/java''. Working directory: /home/l0drex/.gradle/caches/modules-2/files-2.1/com.jetbrains.intellij.idea/ideaIC/2022.1/c91a19824ffcb28872ecd2761c83b250f703ff21/ideaIC-2022.1/bin Command: /home/l0drex/.gradle/caches/modules-2/files-2.1/com.jetbrains/jbre/jbr_jcef-11_0_14_1-linux-x64-b2043.25/extracted/jbr/bin/java -Didea.auto.reload.plugins=true -Didea.classpath.index.enabled=false -Didea.config.path=/home/l0drex/Projekte/auto-dark-mode/plugin/build/idea-sandbox/config -Didea.is.internal=true -Didea.platform.prefix=Idea -Didea.plugins.path=/home/l0drex/Projekte/auto-dark-mode/plugin/build/idea-sandbox/plugins -Didea.required.plugins.id=com.github.weisj.darkmode -Didea.system.path=/home/l0drex/Projekte/auto-dark-mode/plugin/build/idea-sandbox/system -Djava.system.class.loader=com.intellij.util.lang.PathClassLoader -Dsun.awt.disablegrab=true -Xms256m -Xmx512m -Dfile.encoding=UTF-8 -Duser.country=DE -Duser.language=de -Duser.variant -ea -cp /home/l0drex/.gradle/caches/modules-2/files-2.1/com.jetbrains.intellij.idea/ideaIC/2022.1/c91a19824ffcb28872ecd2761c83b250f703ff21/ideaIC-2022.1/lib/3rd-party-rt.jar:/home/l0drex/.gradle/caches/modules-2/files-2.1/com.jetbrains.intellij.idea/ideaIC/2022.1/c91a19824ffcb28872ecd2761c83b250f703ff21/ideaIC-2022.1/lib/util.jar:/home/l0drex/.gradle/caches/modules-2/files-2.1/com.jetbrains.intellij.idea/ideaIC/2022.1/c91a19824ffcb28872ecd2761c83b250f703ff21/ideaIC-2022.1/lib/util_rt.jar:/home/l0drex/.gradle/caches/modules-2/files-2.1/com.jetbrains.intellij.idea/ideaIC/2022.1/c91a19824ffcb28872ecd2761c83b250f703ff21/ideaIC-2022.1/lib/jna.jar com.intellij.idea.Main
2022-05-12T18:10:35.399+0200 [DEBUG] [org.gradle.process.internal.DefaultExecHandle] Changing state to: STARTING
2022-05-12T18:10:35.399+0200 [DEBUG] [org.gradle.process.internal.DefaultExecHandle] Waiting until process started: command '/home/l0drex/.gradle/caches/modules-2/files-2.1/com.jetbrains/jbre/jbr_jcef-11_0_14_1-linux-x64-b2043.25/extracted/jbr/bin/java'.
2022-05-12T18:10:35.400+0200 [DEBUG] [org.gradle.process.internal.DefaultExecHandle] Changing state to: STARTED
2022-05-12T18:10:35.400+0200 [DEBUG] [org.gradle.process.internal.ExecHandleRunner] waiting until streams are handled...
2022-05-12T18:10:35.400+0200 [INFO] [org.gradle.process.internal.DefaultExecHandle] Successfully started process 'command '/home/l0drex/.gradle/caches/modules-2/files-2.1/com.jetbrains/jbre/jbr_jcef-11_0_14_1-linux-x64-b2043.25/extracted/jbr/bin/java''
2022-05-12T18:10:35.520+0200 [ERROR] [system.err] WARNING: An illegal reflective access operation has occurred
An illegal reflective access operation has occurred

2022-05-12T18:10:35.520+0200 [ERROR] [system.err] WARNING: Illegal reflective access by com.intellij.util.lang.UrlClassLoader (file:/home/l0drex/.gradle/caches/modules-2/files-2.1/com.jetbrains.intellij.idea/ideaIC/2022.1/c91a19824ffcb28872ecd2761c83b250f703ff21/ideaIC-2022.1/lib/util.jar) to field java.lang.ClassLoader.classLoaderValueMap
Illegal reflective access by com.intellij.util.lang.UrlClassLoader (file:/home/l0drex/.gradle/caches/modules-2/files-2.1/com.jetbrains.intellij.idea/ideaIC/2022.1/c91a19824ffcb28872ecd2761c83b250f703ff21/ideaIC-2022.1/lib/util.jar) to field java.lang.ClassLoader.classLoaderValueMap

2022-05-12T18:10:35.520+0200 [ERROR] [system.err] WARNING: Please consider reporting this to the maintainers of com.intellij.util.lang.UrlClassLoader
Please consider reporting this to the maintainers of com.intellij.util.lang.UrlClassLoader

2022-05-12T18:10:35.520+0200 [ERROR] [system.err] WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
Use --illegal-access=warn to enable warnings of further illegal reflective access operations

2022-05-12T18:10:35.520+0200 [ERROR] [system.err] WARNING: All illegal access operations will be denied in a future release
All illegal access operations will be denied in a future release

2022-05-12T18:10:35.827+0200 [ERROR] [system.err] Gtk-Message: 18:10:35.828: Failed to load module "xapp-gtk3-module"
2022-05-12T18:10:35.827+0200 [ERROR] [system.err] Gtk-Message: 18:10:35.828: Failed to load module "canberra-gtk-module"
2022-05-12T18:10:35.827+0200 [ERROR] [system.err] Gtk-Message: 18:10:35.828: Failed to load module "appmenu-gtk-module"
2022-05-12T18:10:36.552+0200 [ERROR] [system.err] 2022-05-12 18:10:36,551 [    876]   WARN - #c.i.i.DebugAttachDetector - Unable to start DebugAttachDetector, please add `--add-exports java.base/jdk.internal.vm=ALL-UNNAMED` to VM options
2022-05-12T18:10:36.852+0200 [ERROR] [system.err] 2022-05-12 18:10:36,852 [   1177]   WARN - #c.i.n.i.NotificationGroupManagerImpl - Notification group CodeWithMe is already registered (group=com.intellij.notification.NotificationGroup@7d6698aa). Plugin descriptor: PluginDescriptor(name=Code With Me, id=com.jetbrains.codeWithMe, descriptorPath=plugin.xml, path=~/.gradle/caches/modules-2/files-2.1/com.jetbrains.intellij.idea/ideaIC/2022.1/c91a19824ffcb28872ecd2761c83b250f703ff21/ideaIC-2022.1/plugins/cwm-plugin, version=221.5080.210, package=null, isBundled=true)
2022-05-12T18:10:38.941+0200 [LIFECYCLE] [org.gradle.cache.internal.DefaultFileLockManager] 
...
2022-05-12T18:12:08.942+0200 [DEBUG] [org.gradle.cache.internal.DefaultFileLockManager] Lock acquired on daemon addresses registry.
2022-05-12T18:12:08.942+0200 [DEBUG] [org.gradle.cache.internal.DefaultFileLockManager] Releasing lock on daemon addresses registry.
2022-05-12T18:12:09.486+0200 [DEBUG] [sun.rmi.transport.tcp] RMI TCP Connection(73)-127.0.0.1: accepted socket from [127.0.0.1:60090]
2022-05-12T18:12:09.486+0200 [DEBUG] [sun.rmi.transport.tcp] RMI TCP Connection(73)-127.0.0.1: (port 40771) op = 80
2022-05-12T18:12:09.486+0200 [DEBUG] [sun.rmi.loader] RMI TCP Connection(73)-127.0.0.1: name = "[Ljava.rmi.server.ObjID;", codebase = "", defaultLoader = jdk.internal.loader.ClassLoaders$PlatformClassLoader@5623f185
2022-05-12T18:12:09.486+0200 [DEBUG] [sun.rmi.loader] RMI TCP Connection(73)-127.0.0.1: name = "java.rmi.dgc.Lease", codebase = "", defaultLoader = jdk.internal.loader.ClassLoaders$PlatformClassLoader@5623f185
2022-05-12T18:12:09.486+0200 [DEBUG] [sun.rmi.loader] RMI TCP Connection(73)-127.0.0.1: name = "java.rmi.dgc.VMID", codebase = "", defaultLoader = jdk.internal.loader.ClassLoaders$PlatformClassLoader@5623f185
2022-05-12T18:12:09.486+0200 [DEBUG] [sun.rmi.loader] RMI TCP Connection(73)-127.0.0.1: name = "[B", codebase = "", defaultLoader = jdk.internal.loader.ClassLoaders$PlatformClassLoader@5623f185
2022-05-12T18:12:09.486+0200 [DEBUG] [sun.rmi.loader] RMI TCP Connection(73)-127.0.0.1: name = "java.rmi.server.UID", codebase = "", defaultLoader = jdk.internal.loader.ClassLoaders$PlatformClassLoader@5623f185
2022-05-12T18:12:18.941+0200 [DEBUG] [org.gradle.cache.internal.DefaultFileLockManager] Waiting to acquire shared lock on daemon addresses registry.
2022-05-12T18:12:18.942+0200 [DEBUG] [org.gradle.cache.internal.DefaultFileLockManager] Lock acquired on daemon addresses registry.
2022-05-12T18:12:18.942+0200 [DEBUG] [org.gradle.cache.internal.DefaultFileLockManager] Releasing lock on daemon addresses registry.
2022-05-12T18:12:18.942+0200 [DEBUG] [org.gradle.cache.internal.DefaultFileLockManager] Waiting to acquire shared lock on daemon addresses registry.
2022-05-12T18:12:18.942+0200 [DEBUG] [org.gradle.cache.internal.DefaultFileLockManager] Lock acquired on daemon addresses registry.
2022-05-12T18:12:18.942+0200 [DEBUG] [org.gradle.cache.internal.DefaultFileLockManager] Releasing lock on daemon addresses registry.
2022-05-12T18:12:20.948+0200 [DEBUG] [sun.rmi.transport.tcp] RMI TCP Connection(70)-127.0.0.1: (port 40771) connection closed
2022-05-12T18:12:20.948+0200 [DEBUG] [sun.rmi.transport.tcp] RMI TCP Connection(70)-127.0.0.1: close connection
2022-05-12T18:12:21.047+0200 [DEBUG] [sun.rmi.transport.tcp] RMI TCP Connection(71)-127.0.0.1: (port 40771) connection closed
2022-05-12T18:12:21.048+0200 [DEBUG] [sun.rmi.transport.tcp] RMI TCP Connection(71)-127.0.0.1: close connection
2022-05-12T18:12:21.121+0200 [DEBUG] [sun.rmi.transport.tcp] RMI TCP Connection(72)-127.0.0.1: (port 40771) connection closed
2022-05-12T18:12:21.121+0200 [DEBUG] [sun.rmi.transport.tcp] RMI TCP Connection(72)-127.0.0.1: close connection
2022-05-12T18:12:24.487+0200 [DEBUG] [sun.rmi.transport.tcp] RMI TCP Connection(73)-127.0.0.1: (port 40771) connection closed
2022-05-12T18:12:24.488+0200 [DEBUG] [sun.rmi.transport.tcp] RMI TCP Connection(73)-127.0.0.1: close connection
2022-05-12T18:12:28.941+0200 [DEBUG] [org.gradle.cache.internal.DefaultFileLockManager] Waiting to acquire shared lock on daemon addresses registry.
2022-05-12T18:12:28.942+0200 [DEBUG] [org.gradle.cache.internal.DefaultFileLockManager] Lock acquired on daemon addresses registry.
2022-05-12T18:12:28.942+0200 [DEBUG] [org.gradle.cache.internal.DefaultFileLockManager] Releasing lock on daemon addresses registry.
2022-05-12T18:12:28.942+0200 [DEBUG] [org.gradle.cache.internal.DefaultFileLockManager] Waiting to acquire shared lock on daemon addresses registry.
2022-05-12T18:12:28.942+0200 [DEBUG] [org.gradle.cache.internal.DefaultFileLockManager] Lock acquired on daemon addresses registry.
2022-05-12T18:12:28.942+0200 [DEBUG] [org.gradle.cache.internal.DefaultFileLockManager] Releasing lock on daemon addresses registry.
...
</pre>
</details>

weisJ commented:
Have you tried commenting out the code which creates the XdgThemeMonitorService?

l0drex commented:
Yes, when I do that it works.
I made some further debugging and this line seems to be the cause:
```kotlin
private val connection = DBusConnection.getConnection(DBusConnection.DBusBusType.SESSION)
```

I don't know why yet, but I did found [another plugin](https://github.com/Hannah-Sten/TeXiFy-IDEA/blob/3464f1bbcb7fcf8bed6e103aaf216addafa6aa5e/src/nl/hannahsten/texifyidea/run/linuxpdfviewer/evince/EvinceConversation.kt#L92) that uses the dbus so that in itself shouldn't be a problem.
I tried to move the connection into the functions instead of using it as a class attribute, but that didn't fix it.

weisJ commented on linux/xdg/src/main/java/com/github/weisj/darkmode/platform/linux/xdg/XdgThemeMonitorService.kt:
```diff
@@ -0,0 +1,37 @@
+import org.freedesktop.dbus.interfaces.DBusSigHandler
+
+class XdgThemeMonitorService : ThemeMonitorService {
+    private val freedesktopConnection = FreedesktopConnection()
+    private val sigHandler = SigHandler()
+    override val isDarkThemeEnabled: Boolean = freedesktopConnection.theme == ThemeMode.DARK
+    override val isSupported: Boolean = freedesktopConnection.theme != ThemeMode.ERROR
```
Try this. Calling out to system resources during class initialization makes debugging deadlocks rather difficult. This way you'll at least know which call is causing the lock (It might not be the creation of the connection but the `theme` call instead)
```suggestion
    private val freedesktopConnection by lazy { FreedesktopConnection() }
    private val sigHandler = SigHandler()
    override val isDarkThemeEnabled: Boolean
        get() = freedesktopConnection.theme == ThemeMode.DARK
    override val isSupported: Boolean
        get() = freedesktopConnection.theme != ThemeMode.ERROR
```

l0drex replied:
I'm pretty sure it's the creation of the connection. When I comment everything out except the creation, the deadlock is still there 😅

weisJ replied:
You can’t comment out `isDarkThemeEnabled` and `isSupported` though as they need to be implemented in order for the class to compile. Did you stub them or what exactly do you mean by commenting out?
Also it would be beneficial to add some unit tests, testing whether it works in isolation outside IntelliJ.

l0drex replied:
Is there a way to run IDE and plugin without the sandbox?

l0drex replied:
> You can’t comment out isDarkThemeEnabled and isSupported though as they need to be implemented in order for the class to compile. Did you stub them or what exactly do you mean by commenting out?
Also, it would be beneficial to add some unit tests, testing whether it works in isolation outside IntelliJ.

Didn't see that, sorry :sweat_smile: 

I just returned true and commented the rest out. I have unittests that I run regularly, but I didn't commit them as I don't know which theme is actually active

l0drex commented:
I have now another weird problem. I wanted to update the `dbus-java` dependency to the latest version in the `libs.versinos.toml` file, but there seems to be some sort of artifact of the old version somewhere.

Specifically, I need to change this:
```kotlin
private val connection = DBusConnection.getConnection(DBusConnection.DBusBusType.SESSION)
```

to this:
```kotlin
private val connection = DBusConnectionBuilder.forSessionBus().build()
```

This works when I test it in a new, empty project, but here it throws an error that `DBusConnection.CONNECTIONS` is private and can't be accessed. This was in the previous version of the library the case.

I checked the dependency menu in the Gradle toolbar, it only shows the correct version of the library. I deleted all build directories, cache directories, `~/.gradle` and `~/.m2` and ran the `clean` task, but the issue persists.

l0drex commented:
Yes, it also happens when I run the build from the command line. Building with the dependency commented out didn't help, I even ran it on a different machine and have the same issue there.

l0drex commented:
To keep you updated, I made a [post in the developer forum](https://intellij-support.jetbrains.com/hc/en-us/community/posts/6106226869010-Can-t-connect-to-session-dbus-on-linux) of JetBrains to see if they can help me with that endless spinning on `runIde`, but they have no idea either.

14mRh4X0r commented:
I took this PR and debugged it; what I have right now (at https://github.com/14mRh4X0r/auto-dark-mode/tree/linux-xdg-desktop) is working for me. I'm not quite sure whether the code style fits, as I'm new to Kotlin, but at least it's usable.

Something I noticed as well though, is that (at least IDEA) already includes a (older) version of dbus-java, so it may be better to use that one. I had to exclude some of its transitive dependencies due to classloading clashes.

The hang seems to be caused by exceptions thrown during initialization.

weisJ commented:
Thank you very much for picking this up.

> I took this PR and debugged it; what I have right now (at [14mRh4X0r/auto-dark-mode@`linux-xdg-desktop`](https://github.com/14mRh4X0r/auto-dark-mode/tree/linux-xdg-desktop?rgh-link-date=2022-12-22T15%3A25%3A25Z)) is working for me. I'm not quite sure whether the code style fits, as I'm new to Kotlin, but at least it's usable. (FWIW: it's @l0drex' code rebased onto master, my fixes are in the last commit.)

I would highly appreciate if you could clean up the commits a bit and submit a PR. We can then hash out the little details.

> Something I noticed as well though, is that (at least IDEA) already includes a (older) version of dbus-java, so it may be better to use that one. I had to exclude some of its transitive dependencies due to classloading clashes.

I think this is a good idea. In this case we could even mark the dependency as `runtimeOnly` to avoid bundling it ourselves.

> The hang seems to be caused by exceptions thrown during initialization.

Good catch!

l0drex commented:
@14mRh4X0r thank you so much for this! I added your changes in several commits (hoping I understood the changes correctly) and added you as a co-author. I also merged the current master to fix the merge conflict.

However, I am not sure how to mark `dbus-java` as a `runtimeOnly` dependency. I naively tried replacing `implementation` with `runtimeOnly` in [this line](https://github.com/l0drex/auto-dark-mode/blob/2d3b8962842c707e14a95cb7b72cf33aa13cbcaf/linux/xdg/build.gradle.kts#L9), but that didn't work.

weisJ commented:
Wonderfull :) 

> However, I am not sure how to mark dbus-java as a runtimeOnly dependency. I naively tried replacing implementation with runtimeOnly in [this line](https://github.com/l0drex/auto-dark-mode/blob/2d3b8962842c707e14a95cb7b72cf33aa13cbcaf/linux/xdg/build.gradle.kts#L9), but that didn't work.

Sorry for the confusion, that was my bad. The dependency would need to be marked as `compileOnly`. Just make sure that the version matches the one packaged with IDEA itself. Also I was wondering whether we could apply this to the coroutines dependency aswell. Is it also already provided by IDEA?

weisJ commented on gradle/libs.versions.toml:
```diff
@@ -10,6 +10,11 @@ darklaf = "3.0.1"
 autoservice-annotations = { module = "com.google.auto.service:auto-service-annotations", version.ref = "autoservice" }
 autoservice-processor = { module = "dev.zacsweers.autoservice:auto-service-ksp", version.ref = "autoserviceKsp" }
 darklaf-nativeUtils = { module = "com.github.weisj:darklaf-native-utils", version.ref = "darklaf" }
+kotlinx-coroutines-core-jvm = {module ="org.jetbrains.kotlinx:kotlinx-coroutines-core-jvm", version ="1.6.1"}
```
This should probably updated to match the kotlin version in `gradle.properties`. Once this is merged I'll work on the properties file the single source of truth for versions.

weisJ commented on linux/xdg/src/main/java/com/github/weisj/darkmode/platform/linux/xdg/FreedesktopConnection.kt:
```diff
@@ -0,0 +1,81 @@
+
+
+class FreedesktopConnection {
+    private val connection: DBusConnection? = try {
+        // Temporarily replace the current tread's contextClassLoader to work around dbus-java's naive service loading
+        val ccl = Thread.currentThread().contextClassLoader
+        Thread.currentThread().contextClassLoader = javaClass.classLoader
+        val conn = DBusConnectionBuilder.forSessionBus().build()
+        Thread.currentThread().contextClassLoader = ccl
+        conn
+    } catch (_: DBusException) {
+        null
+    }
```
There already is a similar "hack" used in [Servi](https://github.com/weisJ/auto-dark-mode/blob/master/base/src/main/java/com/github/weisj/darkmode/platform/ServiceUtil.java). This should probably be extracted into a separate helper method.

l0drex replied:
Should that go into a new class, or is there a utility class where I could add that?

weisJ replied:
There currently isn't a good place to put it. Feel free to create a separate utility class for it.

weisJ commented:
I would also highly appreciate if you could add some tests similar to https://github.com/weisJ/auto-dark-mode/blob/master/plugin/src/test/kotlin/com/github/weisj/darkmode/platform/linux/gtk/GtkNativeTest.kt.

l0drex commented:
I extracted the theme changer so that it can be used by other tests. I also added a theme changer for kde, however getting the currently used theme is not so straight forward. `lookandfeeltool` doesn't provide this, and other tools like `kreadconfig5` seem to no longer be able to do this.

weisJ commented:
> I extracted the theme changer so that it can be used by other tests. I also added a theme changer for kde, however getting the currently used theme is not so straight forward. `lookandfeeltool` doesn't provide this, and other tools like `kreadconfig5` seem to no longer be able to do this.

I'm not sure myself how to tackle this currently. I think it's ok to leave this as a FIXME for now, as it doesn't influence the tests that much beside the fact that we have to trust the theme being really changed (and if run locally the theme not being preserved for the developer)

weisJ commented:
Tests are currently failing because there is no xserver running causing the test to be executed headless. I had a similar issue in one of my other projects and the following seemed to fix it:
https://github.com/weisJ/darklaf/blob/e2d1e56c24f147737fb9ce173f3362efac429437/.github/workflows/gradle.yml#L65

weisJ commented:
Sorry for not getting back any sooner. The test not being able to run is rather unfortunate, but I don't have a solution for that right now. I guess we would have to guarantee that the action runner actuallyhas a working dbus with the correct properties exposed. For now I'll disable the test and provide the patch as an optional preview-version.
